pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,465 @@
+	defer klog.InfoS("Shutting down controller", "controller", controllerName)
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem
+// function in order to read and process a message on the workqueue.
+func (c *MCRouteController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *MCRouteController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if k, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.InfoS("Expected string in work queue but got", "object", obj)
+		return true
+	} else if err := c.syncMCFlows(); err == nil {
+		c.queue.Forget(k)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(k)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+	}
+	return true
+}
+
+func (c *MCRouteController) syncMCFlows() error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing flows for Multi-cluster", "time", time.Since(startTime))
+	}()
+	var activeGW *mcv1alpha1.Gateway
+	gws, err := c.gwLister.Gateways(c.namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	activeGW = c.getActiveGateway(gws)
+	if activeGW == nil && c.installedActiveGWName == "" {
+		klog.V(2).InfoS("No active Gateway is found")
+		return nil
+	}
+
+	// Active Gateway Name doesn't change but do a fully flows resync
+	// for any Gateway Spec or ClusterInfoImport changes.
+	klog.InfoS("installed GW", "gateway", c.installedActiveGWName)
+	if activeGW != nil && activeGW.Name == c.installedActiveGWName {
+		if err := c.syncInterClusterFlows(activeGW); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if c.installedActiveGWName != "" {
+		if err := c.cleanupFlows(c.installedActiveGWName); err != nil {
+			return err
+		}
+	}
+
+	if activeGW != nil {
+		if activeGW.Name == c.nodeConfig.Name {
+			err = c.installFlowsOnGateway(activeGW)
+		} else {
+			err = c.installFlowsOnRegularNode(activeGW)
+		}
+		if err != nil {
+			return err
+		}
+		c.installedActiveGWName = activeGW.Name
+	} else {
+		c.installedActiveGWName = ""
+	}
+	return nil
+}
+
+func (c *MCRouteController) syncInterClusterFlows(activeGW *mcv1alpha1.Gateway) error {
+	var installedCIImports []*mcv1alpha1.ClusterInfoImport
+	installedCIImportObjects := c.installedCIImports.List()
+	for _, ciImp := range installedCIImportObjects {
+		installedCIImports = append(installedCIImports, ciImp.(*mcv1alpha1.ClusterInfoImport))
+	}
+	desiredCIImports, err := c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, ciImport := range desiredCIImports {
+		if err := c.addMCFlowsForCIImps(activeGW, ciImport); err != nil {
```

Comment:
We can check if CIImport already installed or not, and if yes skip the call?

Reply:
CIImport may also change, so I do a full sync here as well considering both Gateway and CIImport events are infrequent, otherwise, we need to compare each CIImport in both desiredCIImports and installedCIImports.

Reply:
Then why not just call addMCFlowsForCIImps()? 

But if we assume CIImports can change, we need to remove old flows first? Should we cache CIImports and compare, which might be simpler and more efficient?

Reply:
oh, yes, I think we can call addMCFlowsForCIImps() here, CIImports change can be a cache update, no need to clean and reintall it again since cache key is the same unless the CIImport name or cluster ID are changed, but it usually means a new one instead of an update.

Reply:
The problem is the generated flow might change if the CIImport contents change (e.g. ServiceCIDR changes), and then the generated flow will not override the previous flows in OVS. Cache key is only for updating cache, not the real flows in OVS.

Reply:
Ok. I see flow installation will auto clean-up all flows with the same key. But I still like to understand do we recommend to modify flows even no change. For example, will that reset flow stats? @wenyingd ?